src/libraries/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Unix.cs

```diff
@@ -7,70 +7,135 @@
+        // extracted from the Socket's SafeHandle.
         // This allows operations related to file descriptors to be performed directly on the SafePipeHandle,
-        // and operations that should go through the Socket to be done via _namedPipeSocket.  We keep the
+        // and operations that should go through the Socket to be done via PipeSocket. We keep the
         // Socket's SafeHandle alive as long as this SafeHandle is alive.
 
-        private Socket? _namedPipeSocket;
-        private SafeHandle? _namedPipeSocketHandle;
+        private Socket? _pipeSocket;
+        private SafeHandle? _pipeSocketHandle;
+        private volatile int _disposed;
 
         internal SafePipeHandle(Socket namedPipeSocket) : base(ownsHandle: true)
         {
-            Debug.Assert(namedPipeSocket != null);
-            _namedPipeSocket = namedPipeSocket;
-
-            _namedPipeSocketHandle = namedPipeSocket.SafeHandle;
-
-            bool ignored = false;
-            _namedPipeSocketHandle.DangerousAddRef(ref ignored);
-            SetHandle(_namedPipeSocketHandle.DangerousGetHandle());
+            SetPipeSocketInterlocked(namedPipeSocket, ownsHandle: true);
+            base.SetHandle(_pipeSocketHandle!.DangerousGetHandle());
         }
 
-        internal Socket? NamedPipeSocket => _namedPipeSocket;
-        internal SafeHandle? NamedPipeSocketHandle => _namedPipeSocketHandle;
+        internal Socket PipeSocket => _pipeSocket ?? CreatePipeSocket();
+
+        internal SafeHandle? PipeSocketHandle => _pipeSocketHandle;
 
         protected override void Dispose(bool disposing)
         {
             base.Dispose(disposing); // must be called before trying to Dispose the socket
-            if (disposing && _namedPipeSocket != null)
+            _disposed = 1;
+            Socket? socket;
+            if (disposing && (socket = Volatile.Read(ref _pipeSocket)) != null)
             {
-                _namedPipeSocket.Dispose();
-                _namedPipeSocket = null;
+                socket.Dispose();
+                _pipeSocket = null;
             }
         }
 
         protected override bool ReleaseHandle()
         {
             Debug.Assert(!IsInvalid);
 
-            // Clean up resources for named handles
-            if (_namedPipeSocketHandle != null)
+            if (_pipeSocketHandle != null)
             {
-                SetHandle(DefaultInvalidHandle);
-                _namedPipeSocketHandle.DangerousRelease();
-                _namedPipeSocketHandle = null;
+                base.SetHandle((IntPtr)DefaultInvalidHandle);
+                _pipeSocketHandle.DangerousRelease();
+                _pipeSocketHandle = null;
                 return true;
             }
-
-            // Clean up resources for anonymous handles
-            return (long)handle >= 0 ?
-                Interop.Sys.Close(handle) == 0 :
-                true;
+            else
```

Comment:
In what situation do we hit the else block?

Reply:
When nothing happened to the `SafePipeHandle` that caused the `_pipeSocketHandle
` to be created.

Reply:
I see, we lazily allocate it in some cases via the PipeSocket property. Ok, thanks.